include/swift/AST/ExtInfo.h

```diff
@@ -172,6 +172,14 @@ enum class SILFunctionTypeRepresentation : uint8_t {
   /// constructor). Except for
   /// handling the "this" argument, has the same behavior as "CFunctionPointer".
   CXXMethod,
+
+  /// A KeyPath accessor function, which is thin and also uses the variadic
+  /// length
+  /// generic components serialization in trailing buffer.
+  KeyPathAccessorGetter,
+  KeyPathAccessorSetter,
+  KeyPathAccessorEquals,
+  KeyPathAccessorHash,
```

Comment:
I don't think it's the ideal design to have conventions for each signature, but I couldn't come up with better ideas...

Currently, accessors are represented by the following in SIL:
```swift
sil hidden @testGetter : $@convention(keypath_accessor_getter) (@in_guaranteed S, @in_guaranteed (S, C)) -> @out Int
sil hidden @testSetter : $@convention(keypath_accessor_setter) (@in_guaranteed Int, @in_guaranteed S, @in_guaranteed (S, C)) -> ()
sil hidden @testEq     : $@convention(keypath_accessor_equals) (@in_guaranteed (S, C), @in_guaranteed (S, C)) -> Bool
sil hidden @testHash   : $@convention(keypath_accessor_hash)   (@in_guaranteed (S, C)) -> Int
```

As a possible idea, if having a single `@convention(keypath_accessor)` and put new ParameterConvention attributes like `@keypath_argument_buffer`:

```swift
sil hidden @testGetter : $@convention(keypath_accessor) (@in_guaranteed S, @keypath_argument_buffer (S, C)) -> @out Int
sil hidden @testSetter : $@convention(keypath_accessor) (@in_guaranteed Int, @in_guaranteed S, @keypath_argument_buffer (S, C)) -> ()
sil hidden @testEq     : $@convention(keypath_accessor) (@keypath_argument_buffer (S, C), @keypath_argument_buffer (S, C)) -> Bool
sil hidden @testHash   : $@convention(keypath_accessor) (@keypath_argument_buffer (S, C)) -> Int
```

But unlike other ParameterConventions, `@keypath_argument_buffer` is not only how the parameter should be passed but also a marker to determine where generic parameters should be retrieved from. So I'm unsure whether this new `@keypath_argument_buffer` matches the ParameterConvention's concept or not.

What do you think about this, and do you have any other ideas? It would be much appreciated if I can have any comment from @jckarter :pray:

Reply:
Thank you for your comment 🙏 
Oh, I overlooked `@convention(method)`! It makes sense to me to have such implicit property. I'll update the patch.

Reply:
Looking a bit longer term, for partial apply elimination, I want to extend `SILBoxType` to have a "nonescaping" flavor to explicitl represent the context parameter for nonescaping closures. These have the same need to capture and rebind the generic environment from a parameter (as represented by the `@capturesGenericEnvironment` attribute), so maybe once that is implemented we could also use it for keypath accessors with the normal `thin` convention.